### PR TITLE
LKM: Merge one blackhole into another.

### PR DESCRIPTION
**Pull request by gjz010**
Merged fs and sfs into main project.
Only tested on x86, may even fail to compile on other platforms.
Filesystem refactored.

----

I will try to merge [on my own](https://github.com/gjz010/rcore_plus/pull/3) first.

### CODE REVIEW

File: kernel/.idea/encodings.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jiegec) [inline kernel/.idea/encodings.xml]: we don't need these files.

File: kernel/Cargo.toml [hunk 2]
```diff
@@ -62,9 +62,12 @@ lazy_static = { version = "1.3", features = ["spin_no_std"] }
 smoltcp = { git = "https://github.com/rcore-os/smoltcp", default-features = false, features = ["alloc", "log", "proto-ipv4", "proto-igmp", "socket-icmp", "socket-udp", "socket-tcp", "socket-raw"] }
 bitmap-allocator = { git = "https://github.com/rcore-os/bitmap-allocator" }
 rcore-memory = { path = "../crate/memory" }
+
 rcore-thread = { git = "https://github.com/rcore-os/rcore-thread" }
-rcore-fs = { git = "https://github.com/rcore-os/rcore-fs" }
-rcore-fs-sfs = { git = "https://github.com/rcore-os/rcore-fs" }
+# Merged rcore-fs into kernel. A monolithic kernel is enough.
```
Comment (jiegec) [inline kernel/Cargo.toml:72]: Hmm, we would like separated crates in fact.

File: kernel/src/arch/x86_64/memory.rs [hunk 3]
```diff
@@ -11,11 +11,21 @@ use rcore_memory::paging::*;
 use rcore_memory::PAGE_SIZE;
 
 pub fn init(boot_info: &BootInfo) {
+    //panic!("Crash here");
+    println!("memory::init");
```
Comment (jiegec) [inline kernel/src/arch/x86_64/memory.rs]: Debug statements should be removed.

**Comment by jiegec** [comment]
some files should not be checked into the tree.

**Comment by jiegec** [comment]
So many comments are not removed.

**Comment by jiegec** [comment]
Documents and code formatting are required.

**Comment by jiegec** [comment]
What's the usage of a new vmalloc? The global allocator is implemented for this purpose.

**Comment by jiegec** [comment]
If possible, please rebase upon dev branch.

**Comment by jiegec** [comment]
Please spilt one big commit into small ones.

File: kernel/Cargo.toml (hunk 2 above)
Comment (wangrunji0408) [reply kernel/Cargo.toml:72]: +1

**Comment by wangrunji0408** [comment]
I think we can separate these changes into multiple PRs:
* LKM infrastructure: minimal code which can load and run `hello.ko`
* File system refactors: 

  I prefer to make `VirtualFS` an individual FS implementation. Like: 
  ```rust
  impl FileSystem for VirtualFS {...}
  impl INode for INodeContainer {...}
  ```
  And put it to `rcore-fs-rootfs` in [rcore-fs](https://github.com/rcore-os/rcore-fs/) crate :)

* File system extern "C" interface for LKM: 

  I prefer to separate it to a crate like [rcore-fs-ucore](https://github.com/rcore-os/rcore-fs/blob/master/rcore-fs-ucore/src/lib.rs).

* IPI on x86_64

I'm happy to help (merge the code) if you need. @gjz010

**Comment by gjz010** [comment]
I will do this.

> 在这个分支上整理、合并你这学期的工作，要求每个commit的内容逻辑上相对完整（如增加syscall、加入新的实现代码、修正已有的bug等），在commit log中进行相应的说明并以Signed-off-by的形式在commit log最后留下你的姓名和邮箱（可以用git config设置user.name和user.email后在commit时加-s选项）。为了简化开发过程而进行的非通用修改（如hard code编译的TOOLCHAIN_PREFIX等）不应出现在任何一个commit中

**Pull request closed without merging**